wallet/createtx.go

```diff
@@ -1,125 +1,87 @@
 
-	// A best case tx input serialization cost is 32 bytes of sha, 4 bytes
-	// of output index, 4 bytes of sequnce, and the estimated signature
-	// script size.
-	txInEstimate = 32 + 4 + 4 + sigScriptEstimate
+func makeInputSource(eligible []wtxmgr.Credit) txauthor.InputSource {
+	// Pick largest outputs first.  This is only done for compatibility with
+	// previous tx creation code, not because it's a good idea.
+	sort.Sort(sort.Reverse(byAmount(eligible)))
 
-	// A P2PKH pkScript contains the following bytes:
-	//  - OP_DUP
-	//  - OP_HASH160
-	//  - OP_DATA_20 + 20 bytes of pubkey hash
-	//  - OP_EQUALVERIFY
-	//  - OP_CHECKSIG
-	pkScriptEstimate = 1 + 1 + 1 + 20 + 1 + 1
+	// Current inputs and their total value.  These are closed over by the
+	// returned input source and reused across multiple calls.
+	currentTotal := btcutil.Amount(0)
+	currentInputs := make([]*wire.TxIn, 0, len(eligible))
+	currentScripts := make([][]byte, 0, len(eligible))
 
-	// A best case tx output serialization cost is 8 bytes of value, one
-	// byte of varint, and the pkScript size.
-	txOutEstimate = 8 + 1 + pkScriptEstimate
-)
-
-func estimateTxSize(numInputs, numOutputs int) int {
-	return txOverheadEstimate + txInEstimate*numInputs + txOutEstimate*numOutputs
-}
-
-func feeForSize(incr btcutil.Amount, sz int) btcutil.Amount {
-	return btcutil.Amount(1+sz/1000) * incr
+	return func(target btcutil.Amount) (btcutil.Amount, []*wire.TxIn, [][]byte, error) {
+		for currentTotal < target && len(eligible) != 0 {
+			nextCredit := &eligible[0]
+			eligible = eligible[1:]
+			nextInput := wire.NewTxIn(&nextCredit.OutPoint, nil)
+			currentTotal += nextCredit.Amount
+			currentInputs = append(currentInputs, nextInput)
+			currentScripts = append(currentScripts, nextCredit.PkScript)
+		}
+		return currentTotal, currentInputs, currentScripts, nil
+	}
 }
 
-// InsufficientFundsError represents an error where there are not enough
-// funds from unspent tx outputs for a wallet to create a transaction.
-// This may be caused by not enough inputs for all of the desired total
-// transaction output amount, or due to
-type InsufficientFundsError struct {
-	in, out, fee btcutil.Amount
+// secretSource is an implementation of txauthor.SecretSource for the wallet's
+// address manager.
+type secretSource struct {
+	*waddrmgr.Manager
 }
 
-// Error satisifies the builtin error interface.
-func (e InsufficientFundsError) Error() string {
-	total := e.out + e.fee
-	if e.fee == 0 {
-		return fmt.Sprintf("insufficient funds: transaction requires "+
-			"%s input but only %v spendable", total, e.in)
+func (s secretSource) GetKey(addr btcutil.Address) (*btcec.PrivateKey, bool, error) {
+	ma, err := s.Address(addr)
+	if err != nil {
+		return nil, false, err
 	}
-	return fmt.Sprintf("insufficient funds: transaction requires %s input "+
-		"(%v output + %v fee) but only %v spendable", total, e.out,
-		e.fee, e.in)
+	privKey, err := ma.(waddrmgr.ManagedPubKeyAddress).PrivKey()
```

Comment:
I'll fix but it seems unlikely, and would be a bug in txscript (which calls these functions through the SecretSource interface).